crates/weaver_forge/src/v2/registry.rs

```diff
@@ -0,0 +1,380 @@
+#[serde(deny_unknown_fields)]
+pub struct Signals {
+    /// The metric signals defined.
+    pub metrics: Vec<Metric>,
+    /// The span signals defined.
+    pub spans: Vec<Span>,
+    /// The event signals defined.
+    pub events: Vec<Event>,
+    /// The entity signals defined.
+    pub entities: Vec<Entity>,
+}
+
+/// The set of all refinements for a semantic convention registry.
+///
+/// A refinement is a specialization of a signal for a particular purpose,
+/// e.g. creating a MySQL specific instance of a database span for the purpose
+/// of codegeneration for MySQL.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, JsonSchema)]
+#[serde(deny_unknown_fields)]
+pub struct Refinements {
+    /// The metric refinements defined.
+    pub metrics: Vec<MetricRefinement>,
+    /// The span refinements defined.
+    pub spans: Vec<SpanRefinement>,
+    /// The event refinements defined.
+    pub events: Vec<EventRefinement>,
+}
+
+impl ResolvedRegistry {
+    /// Create a new template registry from a resolved schema registry.
+    pub fn try_from_resolved_schema(
+        schema: weaver_resolved_schema::v2::ResolvedTelemetrySchema,
+    ) -> Result<Self, Error> {
+        let mut errors = Vec::new();
+        let mut attributes: Vec<Attribute> = 
+            schema.catalog.attributes()
+            .iter()
+            .map(|a| Attribute {
+                key: a.key.clone(),
+                r#type: a.r#type.clone(),
+                examples: a.examples.clone(),
+                common: a.common.clone(),
+            })
+            .collect();
+        attributes.sort_by(|l,r| l.key.cmp(&r.key));
+
+        let mut metrics = Vec::new();
+        for metric in schema.registry.metrics {
+            let attributes = metric
+                .attributes
+                .iter()
+                .filter_map(|ar| {
+                    let attr = schema.catalog.attribute(&ar.base).map(|a| MetricAttribute {
+                        base: Attribute {
+                            key: a.key.clone(),
+                            r#type: a.r#type.clone(),
+                            examples: a.examples.clone(),
+                            common: a.common.clone(),
+                        },
+                        requirement_level: ar.requirement_level.clone(),
+                    });
+                    if attr.is_none() {
+                        errors.push(Error::AttributeNotFound {
+                            group_id: format!("metric.{}", &metric.name),
+                            attr_ref: AttributeRef(ar.base.0),
+                        });
+                    }
+                    attr
+                })
+                .collect();
+            metrics.push(Metric {
+                name: metric.name,
+                instrument: metric.instrument,
+                unit: metric.unit,
+                attributes,
+                entity_associations: metric.entity_associations,
```

Comment:
not blocking, we can follow-up: we're checking that attributes exist, but not checking it for entities, it seems inconsistent.